omnigibson/object_states/attached_to.py

```diff
@@ -0,0 +1,293 @@
+m = create_module_macros(module_path=__file__)
+
+m.ATTACHMENT_LINK_PREFIX = "attachment"
+
+m.DEFAULT_POSITION_THRESHOLD = 0.05  # 5cm
+m.DEFAULT_ORIENTATION_THRESHOLD = np.deg2rad(5.0)  # 5 degrees
+m.DEFAULT_JOINT_TYPE = JointType.JOINT_FIXED
+m.DEFAULT_BREAK_FORCE = 10000  # Newton
+m.DEFAULT_BREAK_TORQUE = 10000  # Newton-Meter
+
+
+class AttachedTo(RelativeObjectState, BooleanState, ContactSubscribedStateMixin, JointBreakSubscribedStateMixin, LinkBasedStateMixin):
+    """
+        Handles attachment between two rigid objects, by creating a fixed/spherical joint between self.obj (child) and
+        other (parent). At any given moment, an object can only be attached to at most one other object, i.e.
+        a parent can have multiple children, but a child can only have one parent. Note that only
+        child.states[AttachedTo].get_value(parent) will return True.
+
+        Subclasses ContactSubscribedStateMixin, JointBreakSubscribedStateMixin
+        on_contact function attempts to attach self.obj to other when a CONTACT_FOUND event happens
+        on_joint_break function breaks the current attachment
+    """
+
+    @classproperty
+    def metalink_prefix(cls):
+        """
+        Returns:
+            str: Unique keyword that defines the metalink associated with this object state
+        """
+        return m.ATTACHMENT_LINK_PREFIX
+
+    @staticmethod
+    def get_dependencies():
+        return RelativeObjectState.get_dependencies() + [ContactBodies]
+
+    def _initialize(self):
+        super()._initialize()
+        self.initialize_link_mixin()
+
+        # Reference to the parent object (DatasetObject)
+        self.parent = None
+
+        # Reference to the female meta link of the parent object (RigidPrim)
+        self.parent_link = None
+
+        # Mapping from the female meta link names of self.obj to their children (Dict[str, Optional[DatasetObject] = None])
+        self.children = {link_name: None for link_name in self.links if link_name.split("_")[1].endswith("F")}
+
+        # Cache of parent link candidates for other objects (Dict[DatasetObject, Dict[str, str]])
+        # Map from the male meta link names of self.obj to the correspounding female meta link names of other
+        self.parent_link_candidates = dict()
+
+    def on_joint_break(self, joint_prim_path):
+        # Note that when this function is invoked when a joint break event happens, @self.obj is the parent of the
+        # attachment joint, not the child. We access the child of the broken joint, and call the setter with False
+        child = self.children[joint_prim_path.split("/")[-2]]
+        child.states[AttachedTo].set_value(self.obj, False)
+
+    # Attempts to attach two objects when a CONTACT_FOUND event happens
+    def on_contact(self, other, contact_headers, contact_data):
+        for contact_header in contact_headers:
+            if contact_header.type == ContactEventType.CONTACT_FOUND:
+                self.set_value(other, True)
```

Comment:
hmmm...this could be slightly problematic though, right? if an object happens to touch two other objects simulatenously, but the ordering the contact headers is such that the first object in the queue is not attachable, but the second object is attachable, the object won't be attached to the second object right? maybe the logic should be something like:

```
if self.set_value(other, True):
    break
```

so we keep trying until we find a successful attachment

Reply:
done!